uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/AbstractModuleService.java

```diff
@@ -0,0 +1,256 @@
+                       final String comment) {
+        try {
+            final org.uberfire.java.nio.file.Path projectDirectory = Paths.convert(pathToPomXML).getParent();
+            final Module module2Delete = resourceResolver.resolveModule(Paths.convert(projectDirectory));
+
+            final org.uberfire.java.nio.file.Path parentPom = projectDirectory.getParent().resolve(POM_PATH);
+            POM parent = null;
+            if (ioService.exists(parentPom)) {
+                parent = pomService.load(Paths.convert(parentPom));
+            }
+
+            ioService.delete(projectDirectory,
+                             StandardDeleteOption.NON_EMPTY_DIRECTORIES,
+                             commentedOptionFactory.makeCommentedOption(comment));
+            //Note we do *not* raise a DeleteModuleEvent here, as that is handled by DeleteModuleObserverBridge
+
+            if (parent != null) {
+                parent.setPackaging("pom");
+                parent.getModules().remove(module2Delete.getModuleName());
+                pomService.save(Paths.convert(parentPom),
+                                parent,
+                                null,
+                                "Removing child module " + module2Delete.getModuleName());
+            }
+        } catch (final Exception e) {
+            throw ExceptionUtilities.handleException(e);
+        }
+    }
+
+    @Override
+    public void copy(final Path pathToPomXML,
+                     final String newName,
+                     final String comment) {
+        try {
+            final org.uberfire.java.nio.file.Path projectDirectory = Paths.convert(pathToPomXML).getParent();
+            final org.uberfire.java.nio.file.Path newModulePath = projectDirectory.resolveSibling(newName);
+
+            final POM content = pomService.load(pathToPomXML);
+
+            if (newModulePath.equals(projectDirectory)) {
+                return;
+            }
+
+            if (ioService.exists(newModulePath)) {
+                throw new FileAlreadyExistsException(newModulePath.toString());
+            }
+
+            content.setName(newName);
+            final Path newPathToPomXML = Paths.convert(newModulePath.resolve(POM_PATH));
+            try {
+                ioService.startBatch(newModulePath.getFileSystem());
+                ioService.copy(projectDirectory,
+                               newModulePath,
+                               commentedOptionFactory.makeCommentedOption(comment));
+                pomService.save(newPathToPomXML,
+                                content,
+                                null,
+                                comment);
+            } catch (final Exception e) {
+                throw e;
+            } finally {
+                ioService.endBatch();
+            }
+            final Module newModule = resourceResolver.resolveModule(Paths.convert(newModulePath));
+            newModuleEvent.fire(new NewModuleEvent(newModule,
```

Comment:
In this case it's fine, since presumably a failure would throw an exception, which is rethrown in the catch block.

Even in case of failure, we always want to call endBatch to prevent leaving the IOService in an invalid state.

Reply:
My mistake. It appears that we need to fire the rename event before the delete event (or at least that behaviour is explicitly tested for) so I have gone with your proposal and move the event firing into the try block.